pkg/csi/service/common/vsphereutil.go

```diff
@@ -195,6 +195,30 @@ func CreateBlockVolumeUtil(ctx context.Context, clusterFlavor cnstypes.CnsCluste
 				Id: cnsSnapshotID,
 			},
 		}
+
+		// select the compatible datastore for the case of create volume from snapshot
+		// step 1: query the datastore of snapshot. By design, snapshot is always located at the same datastore
+		// as the source volume
+		cnsVolume, err := QueryVolumeByID(ctx, manager.VolumeManager, cnsVolumeID)
+		if err != nil {
+			return nil, logger.LogNewErrorf(log,
+				"failed to query datastore for the snapshot %s with error %+v",
+				spec.ContentSourceSnapshotID, err)
+		}
+
+		// step 2: validate if the snapshot datastore is compatible with datastore candidates in create spec
```

Comment:
@lintongj why is this step needed?

Reply:
CNS CreateVolumeFromSnapshot functionality can only create volume in the same datastore as the snapshot. As we discussed in the design review, we need to check if storage policy is compatible with the snapshot datastore. If not, we should fail the creation.